ctapipe/calib/camera/calibrator.py

```diff
@@ -178,3 +199,32 @@ def __call__(self, event):
         for telid in tel.keys():
             self._calibrate_dl0(event, telid)
             self._calibrate_dl1(event, telid)
+
+
+def shift_waveforms(waveforms, time_shift_samples):
+    """
+    Shift the waveforms by the mean integer shift to mediate
+    time differences between pixels.
+    The remaining shift (mean + fractional part) is returned
+    to be applied later to the extracted peak time.
+    """
+    mean_shift = time_shift_samples.mean()
```

Comment:
Hi @maxnoe,

I miss why you subtract the mean shift:  I would simply shift each waveform for the correspondent integer value, subtracting the mean I would say you introduce a time bias which is event dependent  (because the  time_shift_samples are event dependent). The camera time flat-fielding must be performed afterwards after the remaining-shift correction

Reply:
This is the mean over all pixels to only shift the minimum number of slices to synchronize. This minimizes the number of lost slices at the edges of the window.

Reply:
I see, but it seems to me that you must add it again (e.g. when we correct the pulse time for the camera flat fielding, which is by the way still missing in the calibrator) in order to have the correct  time

Reply:
No, I also added the correction of the peak time.

`time_shift` is the full shift that is required to fully synchronize the pixels. I split this here into the minimal integer part to synchronize the waveforms and the remaining correction. 

The waveforms are then shifted by the integer shift and after extraction the peak time is shifted by the remainder.

See line 178.

Reply:
Ah, ok.

So the idea here is that the dl1 time_shift includes everything: drs4 (event dependent) and flat-field (monitoring dependent) shifts isn't it?

Reply:
yes, this is the event-wise container that needs to be filled by monitoring sources / event source that holds the actual time shift that corresponds to the event.

Reply:
In the future we could imagine that this dl1calib container is updated in multiple steps. The lst event source fills the low-level correction if any is necessary and then the monitoring data updates the values with the slow-changing time flat fielding (maybe interpolated to that specific event if it changes with time)

Reply:
> In the future we could imagine that this dl1calib container is updated in multiple steps. The lst event source fills the low-level correction if any is necessary and then the monitoring data updates the values with the slow-changing time flat fielding (maybe interpolated to that specific event if it changes with time)

Ok, I think this answers the question I made above.

Reply:
> 
> I miss why you subtract the mean shift: I would simply shift each waveform for the correspondent integer value, subtracting the mean I would say you introduce a time bias which is event dependent (because the time_shift_samples are event dependent). The camera time flat-fielding must be performed afterwards after the remaining-shift correction

Hi @FrancaCassol this event-dependency is not relevant for the rest of the analysis, absolute position of pulses in the readout window provides no useful info, only relative pixel timing is important for analysis. So as @maxnoe wrote the approach of minimizing the shifts (hence loss of samples) is reasonable.

Reply:
> In the future we could imagine that this dl1calib container is updated in multiple steps. The lst event source fills the low-level correction if any is necessary and then the monitoring data updates the values with the slow-changing time flat fielding (maybe interpolated to that specific event if it changes with time)

I  agree @maxnoe, it remains to clarify where the work of the event source will be done one day  but it is not for now. I think after this PR we can go on with [https://github.com/cta-observatory/ctapipe_io_lst/pull/44](https://github.com/cta-observatory/ctapipe_io_lst/pull/44)

Reply:
That was my plan!